File: python/src/graph_engineering/durable_json.py
"""Tagged, cross-language Durable JSON encoding and hashing."""

from __future__ import annotations

import math
import re
import struct
from typing import cast

from ._json import normalize_json_string
from .canonical import canonical_sha256
from .models import MAX_SAFE_INTEGER, JsonValue
from .portable_json import PortableJsonError, portable_json_snapshot

_FLOAT_BITS = re.compile(r"^[0-9a-f]{16}$")


class DurableJsonError(ValueError):
    """Raised when tagged Durable JSON is malformed or non-canonical."""


def _encode(value: JsonValue) -> JsonValue:
    if value is None:
        return ["n"]
    if type(value) is bool:
        return ["b", value]
    if type(value) is str:
        return ["s", value]
    if type(value) is int:
        return ["i", value]
    if type(value) is float:
        return ["f", struct.pack(">d", value).hex()]
    if type(value) is list:
        return ["a", [_encode(item) for item in value]]
    if type(value) is dict:
        return [
            "o",
            [[key, _encode(value[key])] for key in sorted(value)],
        ]
    raise AssertionError("portable JSON snapshot returned an unsupported value")


def encode_durable_json(value: object) -> JsonValue:
    """Detach and encode one portable finite JSON value."""

    try:
        snapshot = portable_json_snapshot(value)
    except PortableJsonError as exc:
        raise DurableJsonError(str(exc)) from exc
    return _encode(snapshot)


def _tagged_array(value: object, *, path: str) -> list[JsonValue]:
    if type(value) is not list or not value or type(value[0]) is not str:
        raise DurableJsonError(f"{path}: expected a non-empty tagged array")
    return cast(list[JsonValue], value)


def _decode(value: object, *, path: str) -> JsonValue:
    tagged = _tagged_array(value, path=path)
    tag = tagged[0]
    if tag == "n":
        if len(tagged) != 1:
            raise DurableJsonError(f"{path}: tag 'n' requires arity 1")
        return None
    if tag == "b":
        if len(tagged) != 2 or type(tagged[1]) is not bool:
            raise DurableJsonError(f"{path}: tag 'b' requires one boolean payload")
        return tagged[1]
    if tag == "s":
        if len(tagged) != 2 or type(tagged[1]) is not str:
            raise DurableJsonError(f"{path}: tag 's' requires one string payload")
        return normalize_json_string(tagged[1])
    if tag == "i":
        if (
            len(tagged) != 2
            or type(tagged[1]) is not int
            or abs(tagged[1]) > MAX_SAFE_INTEGER
        ):
            raise DurableJsonError(f"{path}: tag 'i' requires one safe-integer payload")
        return tagged[1]
    if tag == "f":
        bits = tagged[1] if len(tagged) == 2 else None
        if type(bits) is not str or _FLOAT_BITS.fullmatch(bits) is None:
            raise DurableJsonError(f"{path}: tag 'f' requires sixteen lowercase hex digits")
        decoded = float(struct.unpack(">d", bytes.fromhex(bits))[0])
        if not math.isfinite(decoded) or decoded.is_integer():
            raise DurableJsonError(
                f"{path}: tag 'f' must encode a finite non-integer binary64 value"
            )
        return decoded
    if tag == "a":
        payload = tagged[1] if len(tagged) == 2 else None
        if type(payload) is not list:
            raise DurableJsonError(f"{path}: tag 'a' requires one array payload")
        return [_decode(item, path=f"{path}/1/{index}") for index, item in enumerate(payload)]
    if tag == "o":
        payload = tagged[1] if len(tagged) == 2 else None
        if type(payload) is not list:
            raise DurableJsonError(f"{path}: tag 'o' requires one entries-array payload")
        result: dict[str, JsonValue] = {}
        previous: str | None = None
        for index, entry in enumerate(payload):
            if type(entry) is not list or len(entry) != 2 or type(entry[0]) is not str:
                raise DurableJsonError(f"{path}/1/{index}: expected [string, tagged-value]")
            key = normalize_json_string(entry[0])
            if previous is not None and key <= previous:
                raise DurableJsonError(
                    f"{path}/1/{index}/0: object keys must be unique and code-point sorted"
                )
            previous = key
            result[key] = _decode(entry[1], path=f"{path}/1/{index}/1")
        return result
    raise DurableJsonError(f"{path}: unknown Durable JSON tag {tag!r}")


def decode_durable_json(value: object) -> JsonValue:
    """Decode and strictly validate one canonical tagged Durable JSON value."""

    return _decode(value, path="#")


def durable_json_hash(value: object) -> str:
    """Hash the canonical tagged representation of one portable JSON value."""

    return canonical_sha256(encode_durable_json(value))

File: packages/runtime/src/durable-json.ts
import { createHash } from "node:crypto";
import { compareUnicodeCodePoints } from "@graph-engineering/core";
import { snapshotJson } from "./json.js";
import type { JsonValue } from "./types.js";

export type DurableJson =
  | readonly ["n"]
  | readonly ["b", boolean]
  | readonly ["s", string]
  | readonly ["i", number]
  | readonly ["f", string]
  | readonly ["a", readonly DurableJson[]]
  | readonly ["o", readonly (readonly [string, DurableJson])[]];

function freeze<T extends readonly unknown[]>(value: T): T {
  for (const child of value) {
    if (Array.isArray(child)) freeze(child);
  }
  return Object.freeze(value);
}

function floatBits(value: number): string {
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setFloat64(0, value, false);
  return [...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

function bitsFloat(value: string): number {
  const bytes = new Uint8Array(8);
  for (let index = 0; index < 8; index += 1) {
    bytes[index] = Number.parseInt(value.slice(index * 2, index * 2 + 2), 16);
  }
  return new DataView(bytes.buffer).getFloat64(0, false);
}

function encode(value: JsonValue): DurableJson {
  if (value === null) return freeze(["n"] as const);
  if (typeof value === "boolean") return freeze(["b", value] as const);
  if (typeof value === "string") return freeze(["s", value] as const);
  if (typeof value === "number") {
    if (Number.isInteger(value) || Object.is(value, -0)) {
      return freeze(["i", value === 0 ? 0 : value] as const);
    }
    return freeze(["f", floatBits(value)] as const);
  }
  if (Array.isArray(value)) {
    return freeze(["a", Object.freeze(value.map(encode))] as const);
  }
  const record = value as Readonly<Record<string, JsonValue>>;
  const entries = Object.keys(record)
    .sort(compareUnicodeCodePoints)
    .map((key) => freeze([key, encode(record[key] as JsonValue)] as const));
  return freeze(["o", Object.freeze(entries)] as const);
}

/** Encode a detached portable runtime JSON value without losing binary64 bits. */
export function encodeDurableJson(value: unknown): DurableJson {
  return encode(snapshotJson(value));
}

function invalid(message: string): never {
  throw new TypeError(`Invalid Durable JSON: ${message}`);
}

function exactArray(value: JsonValue, length: number, tag?: string): readonly JsonValue[] {
  if (!Array.isArray(value) || value.length !== length) {
    return invalid(tag === undefined ? "expected a tagged array" : `tag '${tag}' has invalid arity`);
  }
  return value;
}

function decode(value: JsonValue): JsonValue {
  const tuple = exactArray(value, Array.isArray(value) ? value.length : 0);
  const tag = tuple[0];
  if (typeof tag !== "string" || tag.length !== 1) return invalid("unknown tag");

  switch (tag) {
    case "n":
      exactArray(tuple, 1, tag);
      return null;
    case "b": {
      const tagged = exactArray(tuple, 2, tag);
      if (typeof tagged[1] !== "boolean") return invalid("tag 'b' requires a boolean");
      return tagged[1];
    }
    case "s": {
      const tagged = exactArray(tuple, 2, tag);
      if (typeof tagged[1] !== "string") return invalid("tag 's' requires a string");
      return tagged[1];
    }
    case "i": {
      const tagged = exactArray(tuple, 2, tag);
      if (typeof tagged[1] !== "number" || !Number.isSafeInteger(tagged[1])) {
        return invalid("tag 'i' requires a safe integer");
      }
      return tagged[1] === 0 ? 0 : tagged[1];
    }
    case "f": {
      const tagged = exactArray(tuple, 2, tag);
      if (typeof tagged[1] !== "string" || !/^[a-f0-9]{16}$/.test(tagged[1])) {
        return invalid("tag 'f' requires sixteen lowercase hexadecimal digits");
      }
      const number = bitsFloat(tagged[1]);
      if (!Number.isFinite(number) || Number.isInteger(number) || Object.is(number, -0)) {
        return invalid("tag 'f' must encode a finite non-integer binary64 value");
      }
      return number;
    }
    case "a": {
      const tagged = exactArray(tuple, 2, tag);
      if (!Array.isArray(tagged[1])) return invalid("tag 'a' requires an item array");
      return Object.freeze(tagged[1].map(decode));
    }
    case "o": {
      const tagged = exactArray(tuple, 2, tag);
      if (!Array.isArray(tagged[1])) return invalid("tag 'o' requires an entry array");
      const result = Object.create(null) as Record<string, JsonValue>;
      let previous: string | undefined;
      for (const entry of tagged[1]) {
        if (!Array.isArray(entry) || entry.length !== 2 || typeof entry[0] !== "string") {
          return invalid("tag 'o' entries must be [string, tagged-value] pairs");
        }
        if (previous !== undefined && compareUnicodeCodePoints(previous, entry[0]) >= 0) {
          return invalid("tag 'o' keys must be unique and sorted by Unicode code point");
        }
        result[entry[0]] = decode(entry[1] as JsonValue);
        previous = entry[0];
      }
      return Object.freeze(result);
    }
    default:
      return invalid(`unknown tag '${tag}'`);
  }
}

/** Decode and strictly validate a canonical tagged Durable JSON value. */
export function decodeDurableJson(value: unknown): JsonValue {
  const snapshot = snapshotJson(value);
  const decoded = decode(snapshot);
  const canonical = encode(decoded);
  if (JSON.stringify(canonical) !== JSON.stringify(snapshot)) {
    return invalid("value is not in canonical tagged form");
  }
  // The decoder builds objects with a null prototype so assigning an own
  // `__proto__` key cannot mutate a prototype. Re-snapshot only after the
  // canonical tagged form has been proven; callers then receive the same
  // deeply frozen, ordinary JSON object shape as every other runtime boundary.
  return snapshotJson(decoded);
}

/** SHA-256 of canonical UTF-8 JSON for the tagged representation of a value. */
export function durableJsonHash(value: unknown): string {
  const tagged = encodeDurableJson(value);
  return createHash("sha256").update(JSON.stringify(tagged), "utf8").digest("hex");
}
